File: 2024/5/main.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

type FormattedData struct {
	rules   []string
	updates []string
}

func readData() ([]string, error) {
	data, err := os.ReadFile(filepath.Join(".", "data.txt"))
	if err != nil {
		return nil, err
	}
	return strings.Split(string(data), "\n"), nil
}

func formatData(rows []string) FormattedData {
	var rules, updates []string
	isEndOfRules := false

	for _, row := range rows {
		if row == "" {
			isEndOfRules = true
			continue
		}
		if !isEndOfRules {
			rules = append(rules, row)
		} else {
			updates = append(updates, row)
		}
	}
	return FormattedData{rules: rules, updates: updates}
}

func part1(data FormattedData) int {
	rulesSet := make(map[string]bool)
	for _, rule := range data.rules {
		rulesSet[rule] = true
	}

	var validUpdates [][]string

	for _, update := range data.updates {
		pages := strings.Split(update, ",")
		isValidUpdate := true

		for i := 0; i < len(pages)-1; i++ {
			failingRule := pages[i+1] + "|" + pages[i]
			if rulesSet[failingRule] {
				isValidUpdate = false
				break
			}
		}
		if isValidUpdate {
			validUpdates = append(validUpdates, pages)
		}
	}

	sum := 0
	for _, validUpdate := range validUpdates {
		mid, _ := strconv.Atoi(validUpdate[len(validUpdate)/2])
		sum += mid
	}

	fmt.Println(sum)
	return sum
}

func part2(data FormattedData) int {
	var correctedUpdates [][]int

	for _, update := range data.updates {
		pages := make([]int, 0)
		for _, p := range strings.Split(update, ",") {
			num, _ := strconv.Atoi(p)
			pages = append(pages, num)
		}

		rulesSet := make([][2]int, 0)
		for _, rule := range data.rules {
			parts := strings.Split(rule, "|")
			a, _ := strconv.Atoi(parts[0])
			b, _ := strconv.Atoi(parts[1])

			containsA := false
			containsB := false
			for _, p := range pages {
				containsA = containsA || p == a
				containsB = containsB || p == b
			}
			if containsA && containsB {
				rulesSet = append(rulesSet, [2]int{a, b})
			}
		}

		dependencies := make(map[int]int)
		for _, rule := range rulesSet {
			dependencies[rule[1]]++
		}

		corrected := make([]int, 0)
		for len(corrected) < len(pages) {
			for _, page := range pages {

				containsPage := false
				for _, c := range corrected {
					containsPage = containsPage || c == page
					if containsPage {
						break
					}
				}
				if containsPage {
					continue
				}

				if dependencies[page] < 1 {
					corrected = append(corrected, page)
					for _, rule := range rulesSet {
						if rule[0] == page {
							dependencies[rule[1]]--
						}
					}
				}
			}
		}

		original := strings.Trim(strings.Join(strings.Split(update, ","), ","), " ")
		correctedStr := ""
		for i, num := range corrected {
			if i > 0 {
				correctedStr += ","
			}
			correctedStr += strconv.Itoa(num)
		}

		if original != correctedStr {
			correctedUpdates = append(correctedUpdates, corrected)
		}
	}

	sum := 0
	for _, correctedUpdate := range correctedUpdates {
		sum += correctedUpdate[len(correctedUpdate)/2]
	}

	fmt.Println(sum)
	return sum
}

func main() {
	reports, err := readData()
	if err != nil {
		fmt.Println("Error:", err)
		return
	}

	formattedData := formatData(reports)
	part1(formattedData)
	part2(formattedData)
}

File: 2024/5/main.rs
use std::fs;
use std::path::Path;
use std::error::Error;
use std::collections::{HashMap, HashSet};
use tokio;

async fn read_data() -> Result<Vec<String>, Box<dyn Error>> {
	let contents = fs::read_to_string(Path::new("data.txt"))?;
	Ok(contents.lines().map(String::from).collect())
}


#[derive(Debug)]
struct FormattedData {
    rules: Vec<String>,
    updates: Vec<String>,
}

fn format_data(rows: Vec<String>) -> FormattedData {
    let mut rules = Vec::new();
    let mut updates = Vec::new();
    let mut is_end_of_rules = false;

    for line in rows.iter() {
        if line.is_empty() {
            is_end_of_rules = true;
            continue;
        }
        if !is_end_of_rules {
            rules.push(line.to_string());
        } else {
            updates.push(line.to_string());
        }
    }

    FormattedData { rules, updates }
}

fn part1(data: &FormattedData) -> i32 {
    let rules_set: HashSet<_> = data.rules.iter().cloned().collect();
    let mut valid_updates = Vec::new();

    for update in &data.updates {
        let pages: Vec<&str> = update.split(',').collect();
        let mut is_valid_update = true;

        for i in 0..pages.len() - 1 {
            let failing_rule = format!("{}|{}", pages[i + 1], pages[i]);
            if rules_set.contains(&failing_rule) {
                is_valid_update = false;
                break;
            }
        }
        if is_valid_update {
            valid_updates.push(pages);
        }
    }

    let mut sum = 0;
    for valid_update in valid_updates {
        sum += valid_update[valid_update.len() / 2].parse::<i32>().unwrap();
    }

    println!("{}", sum);
    sum
}

fn part2(data: &FormattedData) -> i32 {
    let mut corrected_updates = Vec::new();

    for update in &data.updates {
        let pages: Vec<i32> = update.split(',').map(|x| x.parse().unwrap()).collect();

        let mut rules_set = Vec::new();
        for rule in &data.rules {
            let parts: Vec<i32> = rule.split('|').map(|x| x.parse().unwrap()).collect();
            if pages.contains(&parts[0]) && pages.contains(&parts[1]) {
                rules_set.push((parts[0], parts[1]));
            }
        }

        let mut dependencies: HashMap<i32, i32> = HashMap::new();
        for &(_, b) in &rules_set {
            *dependencies.entry(b).or_insert(0) += 1;
        }

        let mut corrected = Vec::new();
        while corrected.len() < pages.len() {
            for &page in &pages {
                if !corrected.contains(&page) {
                    if dependencies.get(&page).unwrap_or(&0) < &1 {
                        corrected.push(page);
                        for &(a, b) in &rules_set {
                            if page == a {
                                if let Some(count) = dependencies.get_mut(&b) {
                                    *count -= 1;
                                }
                            }
                        }
                    }
                }
            }
        }

        let original = pages
            .iter()
            .map(|x| x.to_string())
            .collect::<Vec<String>>()
            .join(",");

        let corrected_str = corrected
            .iter()
            .map(|x| x.to_string())
            .collect::<Vec<String>>()
            .join(",");

        if original != corrected_str {
            corrected_updates.push(corrected);
        }
    }

    let mut sum = 0;
    for corrected_update in corrected_updates {
        sum += corrected_update[corrected_update.len() / 2];
    }

    println!("{}", sum);
    sum
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
	let data = read_data().await?;
	let formatted = format_data(data);
	part1(&formatted);
	part2(&formatted);
	Ok(())
}